parlai/agents/transformer/module_extensions.py

```diff
@@ -0,0 +1,195 @@
+        )
+        group.add_argument(
+            '--r3f-encoder-noise',
+            type=bool,
+            default=True,
+            help='Add noise to encoder. At least one of `r3f-encoder-noise` and `r3f-coder-noise` must be set to True',
+        )
+        group.add_argument(
+            '--r3f-decoder-noise',
+            type=bool,
+            default=False,
+            help='Add noise to decoder. At least one of `r3f-encoder-noise` and `r3f-coder-noise` must be set to True',
+        )
+
+    def __init__(self, opts):
+        self.use_r3f = opts.get('use_r3f')
+        self.eps = opts.get('r3f_eps')
+        self.noise_type = opts.get('r3f_noise_type')
+        if self.noise_type == R3F_NOISE_NORMAL:
+            self.noise_sampler = torch.distributions.normal.Normal(
+                loc=0.0, scale=self.eps
+            )
+        elif self.noise_type == R3F_NOISE_UNIFORM:
+            self.noise_sampler = torch.distributions.uniform.Uniform(
+                low=-self.eps, high=self.eps
+            )
+        self.r3f_lambda = opts.get("r3f_lambda")
+        self.noise_encoder = opts.get("r3f_encoder_noise")
+        self.noise_decoder = opts.get("r3f_decoder_noise")
+
+        if self.use_r3f and not self.noise_encoder and not self.noise_decoder:
+            raise RuntimeError(
+                "R3FContext: Noise must be added to at least one of the encoder or decoder."
+            )
+        self.had_first_pass = (
+            False
+        )  # necessary because nn.Embedding for encoder/decoder on some models are the same. Use this to only make any relevant deep copies once.
+
+    def _calculate_symm_kl(self, noised_logits, input_logits):
+        return (
+            F.kl_div(
+                F.log_softmax(noised_logits, dim=-1, dtype=torch.float32),
+                F.softmax(input_logits, dim=-1, dtype=torch.float32),
+                None,
+                None,
+                "sum",
+            )
+            + F.kl_div(
+                F.log_softmax(input_logits, dim=-1, dtype=torch.float32),
+                F.softmax(noised_logits, dim=-1, dtype=torch.float32),
+                None,
+                None,
+                "sum",
+            )
+        ) / noised_logits.size(0)
+
+    def forward_pass_with_r3f(self, model, *args, **kwargs):
+        """
+        'model' argument here is the `nn.Module` normally responsible for the full
+        forward pass.
+        """
+        if not self.use_r3f:
+            raise RuntimeError(
+                "Trying to use an R3F loss when this has not been allowed."
+            )
+        regular_output = model(*args, **kwargs)
+        with R3FNoiseEmbeddingContextManager(self, model) as model:
+            noised_output = model(*args, **kwargs)
+        symm_kl = self._calculate_symm_kl(noised_output[0], regular_output[0])
+        return regular_output, self.r3f_lambda * symm_kl
+
+
+class R3FNoisedEmbedding(nn.Module):
```

Comment:
I don't think this is a module, is it?

Reply:
Ah yeah, nice catch. Will remove. 

It's a remnant from before I figured out the "deep copy to get rid of shared embedding between encoder/decoder".